### ISSUE TITLE
Doveadm PW Error

### ISSUE THREAD

**Issue by NicoloLazzaroni** [postfixadmin/postfixadmin#430]
Hello,

I updated from PostFixAdmin 2.2.4 to the latest version of PostFixAdmin.
I regenerated the password hash and in the setup.php tried to create a new admin as I could not login with the old one.

It does not give any error in the webpage itself,b ut then when trying to login the admin is not created. Even checked in the database itself.
I've also tried to reset the admin table in the database.

I get this error in the apache2 error log:

```
[Wed Jan 20 21:21:12.147984 2021] [php7:notice] [pid 28600] [client 188.114.103.114:44816] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:21:12.258988 2021] [php7:notice] [pid 28600] [client 188.114.103.114:44816] Failed to read password from /usr/bin/doveadm pw ... stderr: doveconf: Warning: /etc/dovecot/conf.d/10-mail.conf line 4: Global setting mail_plugins won't change the setting inside an earlier filter at /etc/dovecot/dovecot.conf line 32 (if this is intentional, avoid this warning by moving the global setting before /etc/dovecot/dovecot.conf line 32)\n, password: {SHA512-CRYPT}$6$ENCRYPTEDPASSWORDCENSORED.\n , referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:21:12.259179 2021] [php7:error] [pid 28600] [client 188.114.103.114:44816] PHP Fatal error:  Uncaught Exception: /usr/bin/doveadm pw failed, see error log for details in /srv/postfixadmin/functions.inc.php:1057\nStack trace:\n#0 /srv/postfixadmin/functions.inc.php(1272): _pacrypt_dovecot()\n#1 /srv/postfixadmin/model/PFAHandler.php(563): pacrypt()\n#2 /srv/postfixadmin/public/setup.php(465): PFAHandler->save()\n#3 /srv/postfixadmin/public/setup.php(346): create_admin()\n#4 {main}\n  thrown in /srv/postfixadmin/functions.inc.php on line 1057, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:21:17.720726 2021] [php7:notice] [pid 28691] [client 188.114.103.114:48190] PHP Notice:  Undefined index: theme_favicon in /srv/postfixadmin/model/PFASmarty.php on line 73
[Wed Jan 20 21:21:17.721479 2021] [php7:notice] [pid 28691] [client 188.114.103.114:48190] Config::read(): attempt to read undefined config option "database_socket", returning null
[Wed Jan 20 21:21:20.188326 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] PHP Notice:  Undefined index: theme_favicon in /srv/postfixadmin/model/PFASmarty.php on line 73, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:21:20.188900 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:21:20.221153 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] Failed to read password from /usr/bin/doveadm pw ... stderr: doveconf: Warning: /etc/dovecot/conf.d/10-mail.conf line 4: Global setting mail_plugins won't change the setting inside an earlier filter at /etc/dovecot/dovecot.conf line 32 (if this is intentional, avoid this warning by moving the global setting before /etc/dovecot/dovecot.conf line 32)\n, password: {SHA512-CRYPT}$6$ENCRYPTEDPASSWORDCENSORED (verified)\n , referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:21:20.221259 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] Error while trying to call pacrypt(), referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:21:20.221275 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] Exception: /usr/bin/doveadm pw failed, see error log for details in /srv/postfixadmin/functions.inc.php:1057\nStack trace:\n#0 /srv/postfixadmin/functions.inc.php(1272): _pacrypt_dovecot()\n#1 /srv/postfixadmin/model/Login.php(34): pacrypt()\n#2 /srv/postfixadmin/public/login.php(63): Login->login()\n#3 {main}, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:21:20.221296 2021] [php7:notice] [pid 28596] [client 188.114.103.114:49888] PostfixAdmin admin login failed (username: admin@mydomain.tld, ip_address: 188.114.103.114), referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:22:46.254162 2021] [core:info] [pid 30584] [client 45.148.10.61:47474] AH00128: File does not exist: /var/www/html/boaform/admin/formLogin, referer: http://CENSOREDIP:80/admin/login.asp
[Wed Jan 20 21:24:00.521943 2021] [php7:notice] [pid 28595] [client 188.114.103.114:36082] Config::read(): attempt to read undefined config option "database_socket", returning null
[Wed Jan 20 21:24:06.376903 2021] [php7:notice] [pid 28585] [client 188.114.103.114:39398] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:25:35.767388 2021] [php7:notice] [pid 28600] [client 188.114.103.114:43658] Config::read(): attempt to read undefined config option "database_socket", returning null
[Wed Jan 20 21:25:40.152949 2021] [php7:notice] [pid 28691] [client 188.114.103.114:46956] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:25:52.495570 2021] [php7:notice] [pid 28596] [client 188.114.103.114:55520] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/setup.php




[Wed Jan 20 21:27:05.220148 2021] [php7:error] [pid 28596] [client 188.114.103.114:46082] PHP Fatal error:  Uncaught Exception: /usr/bin/doveadm pw failed, see error log for details in /srv/postfixadmin/functions.inc.php:1056\nStack trace:\n#0 /srv/postfixadmin/functions.inc.php(1271): _pacrypt_dovecot()\n#1 /srv/postfixadmin/model/PFAHandler.php(563): pacrypt()\n#2 /srv/postfixadmin/public/setup.php(465): PFAHandler->save()\n#3 /srv/postfixadmin/public/setup.php(346): create_admin()\n#4 {main}\n  thrown in /srv/postfixadmin/functions.inc.php on line 1056, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:27:07.839598 2021] [php7:notice] [pid 30584] [client 188.114.103.114:47746] PHP Notice:  Undefined index: theme_favicon in /srv/postfixadmin/model/PFASmarty.php on line 73, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:27:07.839953 2021] [php7:notice] [pid 30584] [client 188.114.103.114:47746] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/setup.php
[Wed Jan 20 21:27:17.441619 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] PHP Notice:  Undefined index: theme_favicon in /srv/postfixadmin/model/PFASmarty.php on line 73, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:27:17.442039 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] Config::read(): attempt to read undefined config option "database_socket", returning null, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:27:17.465386 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] Failed to read password from /usr/bin/doveadm pw ... stderr: doveconf: Warning: /etc/dovecot/conf.d/10-mail.conf line 4: Global setting mail_plugins won't change the setting inside an earlier filter at /etc/dovecot/dovecot.conf line 32 (if this is intentional, avoid this warning by moving the global setting before /etc/dovecot/dovecot.conf line 32)\nFatal: Missing {scheme} prefix from hash\n, password:  , referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:27:17.465465 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] Error trying to call pacrypt(), referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:27:17.465479 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] Exception: /usr/bin/doveadm pw failed, see error log for details in /srv/postfixadmin/functions.inc.php:1056\nStack trace:\n#0 /srv/postfixadmin/functions.inc.php(1271): _pacrypt_dovecot()\n#1 /srv/postfixadmin/model/Login.php(45): pacrypt()\n#2 /srv/postfixadmin/public/login.php(63): Login->login()\n#3 {main}, referer: https://mail.mydomain.tld/mail/public/login.php
[Wed Jan 20 21:27:17.465529 2021] [php7:notice] [pid 30813] [client 188.114.103.114:53592] PostfixAdmin admin login failed (username: support@mydomain.tld, ip_address: 188.114.103.114), referer: https://mail.mydomain.tldmail/public/login.php
[Wed Jan 20 21:38:21.572146 2021] [php7:notice] [pid 30813] [client 188.114.102.13:11130] PHP Notice:  Undefined index: theme_favicon in /srv/postfixadmin/model/PFASmarty.php on line 73, referer: https://mail.mydomain.tld/mail/public/login.php
```

I tried to replace the broken class in function.php (_pacrypt_dovecot()), with the old one from the old 2.2.4 just to see what would happen and if the problem would solve, but nothing happened.

Seems like he cannot handle doveadm pw response.

(Doveadm pw works, I tried it on console and as you can see from the logs it does output a response)

Encryption used:` $CONF['encrypt'] = 'dovecot:SHA512-CRYPT';`

I've censored the Encrypted Password with ENCRYPTEDPASSWORDCENSORED even if quite useless, the IP with CENSOREDIP and the domain with mail.mydomain.tld and mydomain.tld. As they are the general Apache2 error logs 2 or 3 lines are useless but to be sure they are not related I didn't remove them.

Thank you very much.

Best Regards

**Comment by DavidGoodwin** [comment]
I think you need to fix your dovecot config, so it stops writing on stderr (the `Warning: /etc/dovecot/conf.d/10-mail.conf line 4: Global setting mail_plugins won't change the setting inside an earlier filter at /etc/dovecot/dovecot.conf line 32 (if this is intentional, avoid this warning by moving the global setting before /etc/dovecot/dovecot.conf line 32` bit. )

**Comment by DavidGoodwin** [comment]
The error that happens after regarding `Fatal: Missing {scheme} prefix from hash` is due to Login's call to pacrypt() with two non-matching passwords, which was added to try and prevent a timing attack against the system.

( See : https://github.com/postfixadmin/postfixadmin/blob/e7e1ce9c283b88cc17bafb9785cd6e30d903785e/model/Login.php#L44 )

**Comment by NicoloLazzaroni** [comment]
Thank you, I'll check right into it and let you know.

**Comment by NicoloLazzaroni** [comment]
Hello,

Now it does works, however I get this error when trying to reset my password:
`[Thu Jan 21 13:38:29.457150 2021] [php7:error] [pid 18330] [client 188.114.103.102:35656] PHP Fatal error:  Uncaught Error: Class 'PFASmarty' not found in /srv/postfixadmin/public/users/password-change.php:28\nStack trace:\n#0 {main}\n  thrown in /srv/postfixadmin/public/users/password-change.php on line 28, referer: https://recover.********.eu/users/password-recover.php
`

Thank you very much for your help.

Best Regards

**Comment by DavidGoodwin** [comment]
Er, I thought I fixed that bug with https://github.com/postfixadmin/postfixadmin/commit/dd86dcad30e79d4ebdf316bbf6701da31c8fe923 ?

And that is/was in 3.3.4 wasn't it?

**Comment by DavidGoodwin** [comment]
oh, sorry ... wrong file! same problem/fix.

**Comment by NicoloLazzaroni** [comment]
Hello,

Solved. Thank you very much!

Best Regards

**Comment by NicoloLazzaroni** [comment]
Hello,

Actually there is another small problem.
I'll have to insert the URL to my configuration this time, then I'll censor it.
https://recover.indst.eu/users/password-recover.php

The CSS and Layout are quite buggy. Is it normal?

Thank You
Best Regards

**Comment by NicoloLazzaroni** [comment]
Hello,

Sorry for the spam. I could resolv the themes issue by miself. Here is the code I used to add the header and change the color if you want to sobstitute it:

**password-recover.tpl:**
`<nav class="navbar navbar-default navbar-fixed-top">
    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand" href='main.php'><img id="login_header_logo" src="{$CONF.theme_logo}"
                                                         alt="Logo"/></a>
            {if $CONF.show_header_text==='YES' && $CONF.header_text}
                <h2>{$CONF.header_text}</h2>
            {/if}
        </div>
    </div>
</nav>

<div id="edit_form">
    <form name="frmPassword" method="post" action="">
        <table>
            <tr>
                <th colspan="3">{$PALANG.pPassword_recovery_title}</th>
            </tr>
            <tr>
                <td class="label"><p style="color:black;"><label>{$PALANG.pLogin_username}:</label></p></td>
                <td><input class="flat" type="text" name="fUsername"/></td>
            </tr>
            <tr>
                <td class="label">&nbsp;</td>
                <td colspan="2">
                    <input class="button" type="submit" name="submit" value="{$PALANG.pPassword_recovery_button}"/>
                </td>
            </tr>
        </table>
    </form>
    {literal}
        <script type="text/javascript">
            
        </script>
    {/literal}
</div>`

**password-change.tpl:**
`<nav class="navbar navbar-default navbar-fixed-top">
    <div class="container">
        <div class="navbar-header">
            <a class="navbar-brand" href='main.php'><img id="login_header_logo" src="{$CONF.theme_logo}"
                                                         alt="Logo"/></a>
            {if $CONF.show_header_text==='YES' && $CONF.header_text}
                <h2>{$CONF.header_text}</h2>
            {/if}
        </div>
    </div>
</nav>

<div id="edit_form">
    <form name="mailbox" method="post">
        <table>
            <tr>
                <td colspan="3"><h3>{$PALANG.pPassword_welcome}</h3></td>
            </tr>
            <tr>
                <td>{$PALANG.pLogin_username} :</td>
                <td><input class="flat" type="text" name="fUsername" value="{$tUsername}"/></td>
            </tr>
            <tr>
                <td>{$PALANG.pPassword_password_code} :</td>
                <td><input class="flat" type="text" name="fCode" value="{$tCode}"/></td>
            </tr>
            <tr>
                <td>{$PALANG.pPassword_password} :</td>
                <td><input class="flat" type="password" name="fPassword"/></td>
            </tr>
            <tr>
                <td>{$PALANG.pPassword_password2} :</td>
                <td><input class="flat" type="password" name="fPassword2"/></td>
            </tr>
            <tr>
                <td colspan="2" class="hlp_center"><input class="button" type="submit" name="submit"
                                                          value="{$PALANG.change_password}"/></td>
            </tr>
        </table>
    </form>
</div>
`

Thank you very much for your help with all these issues.

Best Regards

**Comment by DavidGoodwin** [comment]
Thanks for taking the time to figure out how to add the nav bar stuff into those two templates. The fix should be part of postfixadmin 3.3.5 etc whenever that gets released.
